cumulus/polkadot-parachain/src/rpc.rs

```diff
@@ -18,91 +18,82 @@
-	let FullDeps { client, pool, deny_unsafe } = deps;
-
-	module.merge(System::new(client.clone(), pool, deny_unsafe).into_rpc())?;
-	module.merge(TransactionPayment::new(client.clone()).into_rpc())?;
-	module.merge(StateMigration::new(client, backend, deny_unsafe).into_rpc())?;
-
-	Ok(module)
+	fn build_rpc_extensions(
+		_deny_unsafe: DenyUnsafe,
+		_client: Arc<ParachainClient<RuntimeApi>>,
+		_backend: Arc<ParachainBackend>,
+		_pool: Arc<sc_transaction_pool::FullPool<Block, ParachainClient<RuntimeApi>>>,
+	) -> sc_service::error::Result<RpcExtension> {
+		Ok(RpcExtension::new(()))
+	}
 }
 
-/// Instantiate all RPCs we want at the contracts-rococo chain.
-pub fn create_contracts_rococo<C, P>(
-	deps: FullDeps<C, P>,
-) -> Result<RpcExtension, Box<dyn std::error::Error + Send + Sync>>
+pub(crate) struct BuildParachainRpcExtensions<RuntimeApi>(PhantomData<RuntimeApi>);
+
+impl<RuntimeApi>
+	BuildRpcExtensions<
+		ParachainClient<RuntimeApi>,
+		ParachainBackend,
+		sc_transaction_pool::FullPool<Block, ParachainClient<RuntimeApi>>,
+	> for BuildParachainRpcExtensions<RuntimeApi>
 where
-	C: ProvideRuntimeApi<Block>
-		+ sc_client_api::BlockBackend<Block>
-		+ HeaderBackend<Block>
-		+ AuxStore
-		+ HeaderMetadata<Block, Error = BlockChainError>
-		+ Send
-		+ Sync
-		+ 'static,
-	C::Api: substrate_frame_rpc_system::AccountNonceApi<Block, AccountId, Nonce>,
-	C::Api: pallet_transaction_payment_rpc::TransactionPaymentRuntimeApi<Block, Balance>,
-	C::Api: BlockBuilder<Block>,
-	P: TransactionPool + Sync + Send + 'static,
+	RuntimeApi: ConstructNodeRuntimeApi<Block, ParachainClient<RuntimeApi>> + Send + Sync + 'static,
+	RuntimeApi::RuntimeApi: pallet_transaction_payment_rpc::TransactionPaymentRuntimeApi<Block, Balance>
+		+ substrate_frame_rpc_system::AccountNonceApi<Block, AccountId, Nonce>,
 {
-	use pallet_transaction_payment_rpc::{TransactionPayment, TransactionPaymentApiServer};
-	use sc_rpc::dev::{Dev, DevApiServer};
-	use substrate_frame_rpc_system::{System, SystemApiServer};
-
-	let mut module = RpcExtension::new(());
-	let FullDeps { client, pool, deny_unsafe } = deps;
-
-	module.merge(System::new(client.clone(), pool, deny_unsafe).into_rpc())?;
-	module.merge(TransactionPayment::new(client.clone()).into_rpc())?;
-	module.merge(Dev::new(client, deny_unsafe).into_rpc())?;
-
-	Ok(module)
+	fn build_rpc_extensions(
+		deny_unsafe: DenyUnsafe,
+		client: Arc<ParachainClient<RuntimeApi>>,
+		backend: Arc<ParachainBackend>,
+		pool: Arc<sc_transaction_pool::FullPool<Block, ParachainClient<RuntimeApi>>>,
+	) -> sc_service::error::Result<RpcExtension> {
+		let build = || -> Result<RpcExtension, Box<dyn std::error::Error + Send + Sync>> {
+			let mut module = RpcExtension::new(());
+
+			module.merge(System::new(client.clone(), pool, deny_unsafe).into_rpc())?;
+			module.merge(TransactionPayment::new(client.clone()).into_rpc())?;
+			module.merge(StateMigration::new(client.clone(), backend, deny_unsafe).into_rpc())?;
```

Comment:
So this is one example of a future bottleneck: Not all parachains really need this PRC. But, oh well, it is okay for now.

Reply:
@bkchr any thoughts on this ? I don't know what's the performance penalty of adding this RPC. Should we add back the `BuildContractsRococoRpcExtensions` ?

Reply:
I don't think there is any penalty to adding it here. It would be good to investigate what happens if you run this node with a runtime that doesn't have `pallet-transaction-payment-rpc-runtime-api`s exposed, and see if the error is reasonable? 

and once you have some kind of a `metadata-check`, this will be a good addition to it. If you have an issue for metadata-check, please add RPC checks to it as well :)

Reply:
> about tx-payment, It would be good to investigate what happens if you run this node with a runtime that doesn't have `pallet-transaction-payment`s exposed, and see if the error is reasonable?

Makes sense. Will try to do that.

> and once you have some kind of a `metadata-check`, this will be a good addition to it: To check if the RPC's that are exposed are supported by the runtime. If you have an issue for metadata-check, please add RPC checks to it as well :)

I created a more generic issue to keep track of such requirements and added this one to it: https://github.com/paritytech/polkadot-sdk/issues/4966

Reply:
> about tx-payment, It would be good to investigate what happens if you run this node with a runtime that doesn't have `pallet-transaction-payment`s exposed, and see if the error is reasonable?

It will yell at you because you don't have the runtime api available. People should also not call these RPCs anymore. They should directly call the runtime api

Reply:
Tried to run `queryInfo()` on shell and got the following: `1: Unable to query dispatch info.: Execution failed: Other: Exported method TransactionPaymentApi_query_info is not found`